docs/design/crd-proposal-phase1.md

```diff
@@ -0,0 +1,215 @@
+
+The operator will take the SmbShare and SmbSecurityDomain resources as inputs
+and create as many smbd, winbind, etc. backing services as needed. Users
+will have limited input into what backing services the operator will create.
+The operator may, or may not, combine one or more share into a single smbd
+instance.
+
+SmbShare CRs will refer to one or more SmbSecurityDomain resources. The
+SmbSecurityDomain refernce will define the main security properties of the
+smbd that hosts the share. If two SmbShares are defined with different
+SmbSecurityDomains they may not be combined under one smbd.
+
+The listings below are not meant to be entirely complete but they outline
+the genearal direction to make the operaterator a fully-fledged tool to
+create shares that can be consumed by a variety of clients for various use
+cases. Some sections are deliberately marked "TBD" (to be decided) as
+placeholders for more detail that will be needed even in Phase 1.
+
+## SmbSecurityDomain
+
+Spec Options:
+* `mode` - enumerated string "user" or "active-directory" - Indicates that this is an AD joined instance or based on a local mapping of users
+* `users` - mapping - A new subsection pertaining to "user" options - required when mode = "user"
+  * `secret` - string - The name of a secret containing user and group definitions that will be used to create the local users and groups for the service
+* `localName` - string - Optional string used to indicate how this instance will
+  be referred to in an Active Directory
+* `realm` - string - The dns domain name of an Active Directory domain
+* `domains` - list-of-maps - A new subsection pertaining to detailed "active-directory" configuration (optional)
+  * `backend` - enumerated string - the name of a (supported) ID mapping backend
+  * `domainName` - string - the name of the domain being configured
+  * TBD - options to configure the backend, ID range, etc.
+
+
+## SmbShare
+
+Spec Options:
+* `shareName` - string - Optional string giving the share a full SMB compliant name. If unset, the name will be derived by the operator.
+* `storage` - mapping - A new subsection to configure the storage backing the share.  * `pvc` - mapping - A new subsection used to configure PVC backing storage.
+    * `name` - string - The name of a PVC, managed externally to the operator, for the share
+    * `spec` - mapping, embedded pvc spec - An embedded PVC spec that will be used to dynamically create a backing PVC for the share; sharing the lifecycle of the PVC with the share
+  * TBD - Any other more custom storage backends if needed
+* `securityDomain`  - string - The name of the SmbSecurityDomain CR associated with this share
+* `grouping` - string - Optional string that if undefined will be generated by the operator. The grouping is used by the operator as a hint to combine shares under one smbd service, depending on other settings
+* `clustering` - mapping - A new subsection pertaining to clustering smbd instances
```

Comment:
We might want to structure this a little differently, not yet sure how: As mentioned I was identifying 3 aspects of samba's dbwrap backend currently provided by ctdb
1. indication of cluster intactness in the backend (recovery lock file)
2. volatile DBs (and messaging)
3. persistent DBs.

of these, #1 might be useful without clustering, #3 is definitely useful for making config and settings (join...) persistent. But #2 would really for moving to all-active clustering. This would require development in samba though. 

So to summarize, not sure if putting this under "clustering" is entirely sufficient or appropriate.

Reply:
Sure. At the time I was only thinking of non-clustered or clustered. I do think that some sort of subsection would be needed to what kind of deployment (in the general sense, not just the k8s resource) we want and how it is expected to scale. Perhaps if we replace "clustering" with "scaling" as a more generic term? And then the type of scaling might be single instance or clustered or some hybrid...

Reply:
Also, I forgot to note. None of this is discussing what the defaults would be unless explicitly stated. I suspect that in some cases we might prefer to have the more "complex" mode the default but allow some use cases to "opt-out". For example if a clustered mode was the default, a user might want to scale down to one instance for a read only share of limited importance to save on resources, etc.